From: Kemeng Shi <shikemeng@huaweicloud.com>
Subject: mm/compaction: set compact_cached_free_pfn correctly in update_pageblock_skip
Date: Sun, 30 Jul 2023 01:43:51 +0800

We will set skip to page block of block_start_pfn, it's more reasonable to
set compact_cached_free_pfn to page block before the block_start_pfn.

Link: https://lkml.kernel.org/r/20230729174354.2239980-3-shikemeng@huaweicloud.com
Signed-off-by: Kemeng Shi <shikemeng@huaweicloud.com>
Cc: Baolin Wang <baolin.wang@linux.alibaba.com>
Cc: David Hildenbrand <david@redhat.com>
Cc: Kemeng Shi <shikemeng@huawei.com>
Cc: Matthew Wilcox <willy@infradead.org>
Cc: Mel Gorman <mgorman@techsingularity.net>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/compaction.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/mm/compaction.c~mm-compaction-set-compact_cached_free_pfn-correctly-in-update_pageblock_skip
+++ a/mm/compaction.c
@@ -1716,7 +1716,8 @@ static void isolate_freepages(struct com
 
 		/* Update the skip hint if the full pageblock was scanned */
 		if (isolate_start_pfn >= block_end_pfn)
-			update_pageblock_skip(cc, page, block_start_pfn);
+			update_pageblock_skip(cc, page, block_start_pfn -
+					      pageblock_nr_pages);
 
 		/* Are enough freepages isolated? */
 		if (cc->nr_freepages >= cc->nr_migratepages) {
_
